cypher/cypher-docs/src/docs/dev/query-tuning.asciidoc

```diff
@@ -57,6 +90,40 @@ When Cypher is building execution plans, it looks at the schema to see if it can
 These index decisions are only valid until the schema changes, so adding or removing indexes leads to the execution plan cache being flushed.
 
 
+[[cypher-preparser-runtime]]
+=== Cypher runtime
+
+Using the execution plan, the query is executed -- and records returned -- by the query engine, or _runtime_.
+Depending on whether Neo4j Enterprise Edition or Neo4j Community Edition is used, there are three different runtimes available:
+
+Interpreted::
+In this runtime, the operators in the execution plan are chained together in a tree, where each non-leaf operator feeds from one or two child operators.
+The tree thus comprises nested iterators, and the records are streamed in a pipelined manner from the top iterator, which pulls from the next iterator and so on.
+
+Slotted::
+This is very similar to the interpreted runtime, except that there are additional optimizations regarding the way in which the records are streamed through the iterators.
+This results in improvements to both the performance and memory usage of the query.
+In effect, this can be thought of as a the 'faster interpreted' runtime.
+
+[NOTE]
+The slotted runtime is only available in Neo4j Enterprise Edition.
+
+Compiled::
+Algorithms are employed to group intelligently the operators in the execution plan in order to generate new combinations and orders of execution which are optimised for performance and memory usage.
+While this should lead to superior performance in most cases (over both the interpreted and slotted runtimes), it is still under development.
```

Comment:
I made a comment in the 3.1 review about this not supporting all operators or queries. But both that and what is already written here is also true of slotted. We could describe that both compiled and slotted provide performance advantages, and both do not cover all operators and queries, but while compiled has the highest performance advantage, it also has the lowest coverage. So the fallback mechanism in enterprise is to try compiled first, and if it does not support the query, then try slotted, and if that also does not support the query finally fall back to interpreted which supports all queries. In community, only interpreted is considered.

Reply:
👍